models/yang/sonic/sonic-route-map.yang

```diff
@@ -159,17 +159,34 @@ module sonic-route-map {
                 }
 
                 leaf-list match_neighbor {
-                    type inet:ip-address;
+                    type union {
+                        type inet:ip-address;
+                        type leafref {
+                            path "/prt:sonic-port/prt:PORT/prt:PORT_LIST/prt:ifname";
+                        }
+                        type leafref {
+                            path "/spc:sonic-portchannel/spc:PORTCHANNEL/spc:PORTCHANNEL_LIST/spc:name";
+                        }
+                        type leafref {
+                            path "/svlanint:sonic-vlan-interface/svlanint:VLAN_INTERFACE/svlanint:VLAN_INTERFACE_LIST/svlanint:vlanName";
+                        }
+                    }
                     description
-                        "List of IP addresses in the neighbor set";
-                    max-elements 1;
+                        "Match neighbor IP or interface.";
+                    must "(count(current()) = 1)" {
```

Comment:
Why we have chosen leaf-list if it is going to contain only one element ? We should use leaf in that case.
This does not seem correct modelling to me. 

Not sure why it is changed to must expression from max-elements also. Is max-elements not working ? If the idea is to introduce error-app-tag or error-message with must expression, then please note that must expression is costly operation and may impact performance for large table.

Reply:
Partha - We have two options, 1) Dont deviate OC model but put the restriction based on backend capability 2) Deviate OC model based on backend capability, we chose the first one.


We're using must expression to return an error to the user, in general simple must condition is costly or complex one? what's our recommendation? dont use the must condition in the sonic YANG?

Reply:
If max-elements was doing the job for leaf-list, why we would write must expression ? Were not we getting error message for max-elements ? I agree that YANG RFC does not allow user defined error message in YANG for max-elements syntax though.

'must' expression goes through xpath expression evaluation and in general would be costlier than max-elements.

Reply:
'must' expression can't handle all semantic checks or it may not be performant with given expresison, we have recommend using custom validation for such cases.